records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,187 @@
+[celery-batch](https://pypi.org/project/celery-batches/) project which
+exists that might offer us the option to easily achieve this with minimal
+setup.
+
+### Tweaking autovacuum options
+
+There are a few autovacuum configuration that we can tweak in order to delay
+the autovacuum action that triggers a performance slowdown of our database:
+
+* **autovacuum_scale_factor** (default: 0.2): we could use this to delay the vacuum, (suggestion 0.3).
+* **cost_delay**: We could use this to delay by increasing the threshold.
+
+### Manual execution of the vacuum process on the `notifications` table
+
+An option is to keep the autovacuum enabled but trigger a manual vacuum prior
+to traffic time. This would better prepare the database for upcoming load.
+
+Don't run `vacuum FULL`: this will take on most database resources and prevent
+proper operation of the database.
+
+### Disable the autovacuum process on the `notifications` table
+
+To avoid the autovacuum to process the `notifications` table during the daily
+blast of notifications, we could disable the autovacuum to prevent it from
+triggering. This need to be done carefully, as a maximum limit of 4 billion
+transactions can occur in PostgreSQL. The autovacuum cleanup reset this
+transaction number. If this threshold is reached, our database will be severely
+affected.
+
+This is a risky move and certainly not a long/mid term solution.
+
+### Database connections pooling
+
+To maximise the distribution of database connection, we can have a
+pooler that would sit in-between the Notify stack and the AWS RDS
+Aurora Postgres instances.
+
+RDS Proxies would exists for this purpose. Apparently, we could create an
+endpoint each for both readers and the writer. This could co-exist nicely
+with our existing application logic which dispatch database sessions to
+either the readers or writer.
+
+### Tweak the readiness and liveness Kubernetes probe
+
+About these two probes: liveness probe will check if your pod is
+up abnd running; failure to do so will restart it. Readiness will check if
+your pod can receive traffic; failure to do so will detach the pod from the
+load balancer. Both probes will be continually checked throughout the
+pods' life.
+
+At the moment, the liveness probe targets the simple healthcheck that skips
+the database check while the readiness targets the healthcheck with the
+database check.
+
+Because our readiness probe reports that our pods are not ready anymore and
+hence, get detached from the load balancer, which **then cause notifications to
+be lost forever**, we could tweak it on the short term to accommodate our current
+situation and stop the bleeding.
+
+Normally, a delayed database should not be normal. We know in our current
+situation though that the database is still up and running. We'll receive paging
+alerts anyway if it goes down. We should definitely put back the database
+check once the design is fixed.
+
```

Comment:
It's a good call to ease on the readiness check as you're aware a slow response doesn't mean a dead service.